File: javascript/bob/bob.js
var Message = function(message) {
  this.message = message.replace(/[\r\n]/, ' ');
};
  
Message.prototype.is_silence = function() {
  var msg = this.message;
  return (!msg || /^\s*$/.test(msg));
};

Message.prototype.is_a_question = function() {
  return /\?$/.test(this.message)
};

Message.prototype.is_shouting = function() {
  var msg = this.message;
  var upperMsg = msg.toUpperCase();
  var cleanMsg = msg.replace(/[\s.\,\*!@#\$%\^&\(\)\[\]\|\\\/]+/g, '');
  return (msg === upperMsg) && (!/^\d+\??$/.test(cleanMsg));
};

var Bob = function() {
};

Bob.prototype.hey = function(msg) {
  
  var m = new Message(msg);
  if(m.is_silence()) { return "Fine. Be that way!" }
  else if(m.is_shouting()) { return "Woah, chill out!" }
  else if(m.is_a_question()) { return "Sure." }
  else { return "Whatever." }

};

module.exports = Bob;

File: java/bob/src/main/java/Bob.java
public class Bob {

    private final String YELLING_RESPONSE = "Whoa, chill out!";
    private final String DEFAULT_RESPONSE = "Whatever.";
    private final String SILENCE_RESPONSE = "Fine. Be that way!";
    private final String QUESTION_RESPONSE = "Sure.";

    public String hey(String sentence) {
        String trimmedSentence = sentence.trim();
        String cleanedSentence = trimmedSentence.replaceAll("[\\s\\p{Punct}]", "");
        if (trimmedSentence.isEmpty()) {
            return SILENCE_RESPONSE;
        }
        else if (isYelling(cleanedSentence)) {
            return YELLING_RESPONSE;
        }
        else if (isQuestion(sentence)) {
            return QUESTION_RESPONSE;
        }
        return DEFAULT_RESPONSE;
    }

    boolean isYelling(String cleanSentence) {
        return !cleanSentence.matches("^\\d+") && (cleanSentence.equals(cleanSentence.toUpperCase()));
    }
    boolean isQuestion(String sentence) {
        return sentence.charAt(sentence.length() - 1) == '?';
    }
}